core/store/src/trie/mem/arena/alloc.rs

```diff
@@ -0,0 +1,124 @@
+        (size_class + 1) * 8
+    } else if size_class <= allocation_class(ROUND_UP_TO_64_BYTES_UNDER) {
+        (size_class - allocation_class(ROUND_UP_TO_8_BYTES_UNDER)) * 64 + ROUND_UP_TO_8_BYTES_UNDER
+    } else {
+        ROUND_UP_TO_64_BYTES_UNDER << (size_class - allocation_class(ROUND_UP_TO_64_BYTES_UNDER))
+    }
+}
+
+const NUM_ALLOCATION_CLASSES: usize = allocation_class(MAX_ALLOC_SIZE) + 1;
+
+impl Allocator {
+    pub fn new() -> Self {
+        Self { freelists: [usize::MAX; NUM_ALLOCATION_CLASSES], next_ptr: 0 }
+    }
+
+    /// Allocates a slice of the given size in the arena.
+    pub fn allocate<'a>(&mut self, arena: &'a mut ArenaMemory, size: usize) -> ArenaSliceMut<'a> {
+        assert!(size <= MAX_ALLOC_SIZE, "Cannot allocate {} bytes", size);
+        let size_class = allocation_class(size);
+        let allocation_size = allocation_size(size_class);
+        if self.freelists[size_class] == usize::MAX {
+            if arena.mmap.len() < self.next_ptr + allocation_size {
+                panic!("Arena out of memory");
+            }
+            let ptr = self.next_ptr;
+            self.next_ptr += allocation_size;
+            arena.slice_mut(ptr, size)
+        } else {
+            let pos = self.freelists[size_class];
+            self.freelists[size_class] = arena.ptr(pos).read_usize();
+            arena.slice_mut(pos, size)
+        }
+    }
+
+    /// Deallocates the given slice from the arena; the slice's `pos` and `len`
+    /// must be the same as an allocation that was returned earlier.
+    pub fn deallocate(&mut self, arena: &mut ArenaMemory, pos: usize, len: usize) {
+        let size_class = allocation_class(len);
+        arena
+            .slice_mut(pos, allocation_size(size_class))
+            .write_usize_at(0, self.freelists[size_class]);
+        self.freelists[size_class] = pos;
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::trie::mem::arena::Arena;
+
+    use super::MAX_ALLOC_SIZE;
+
+    #[test]
+    fn test_allocate_deallocate() {
+        let mut arena = Arena::new(10000);
+        // Repeatedly allocate and deallocate; we should not run out of memory.
+        for i in 0..1000 {
+            let mut slices = Vec::new();
+            for size in 1..=100 {
+                let mut alloc = arena.alloc(size);
+                // Check that the allocated length is large enough.
+                assert!(alloc.len >= size);
+                let region = alloc.raw_slice_mut();
+                // Try writing some arbitrary bytes into the allocated space.
+                for j in 0..size {
+                    region[j] = ((i + j) % 256) as u8;
+                }
+                slices.push((alloc.pos, alloc.len));
+            }
+            slices.sort_by_key(|(pos, _)| *pos);
+            // Check that the allocated intervals don't overlap.
+            for i in 1..slices.len() {
+                assert!(slices[i - 1].0 + slices[i - 1].1 <= slices[i].0);
+            }
```

Comment:
Also assert that the last allocation has `pos+len < 10000`? It’s probably fine anyway as you’re writing random bytes, but it doesn’t cost much to add an assert